### PR TITLE
[pipeline] build bloom model and policy , revise the base class of policy

### PR DESCRIPTION
**Pull request by CjhHa1**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`
#4119 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
This PR adds bloom model for pipeline forward and policy, revises the base Policy class and adds new feature.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/pipeline/policy/bert.py [hunk 1]
```diff
@@ -254,9 +263,9 @@ def get_hold_layers(self, module: BertModel) -> List[Module]:
         if self.stage_manager.is_first_stage():
             hold_layers.append(module.embeddings)
         num_layers_per_stage_accumulated = np.cumsum(self.layers_per_stage)
-        hold_layers.extend(module.encoder.layer[num_layers_per_stage_accumulated \
-                    [self.stage_manager.stage-1] if self.stage_manager.stage > 0 else 0:
-                    num_layers_per_stage_accumulated[self.stage_manager.stage]])
+        hold_layers.extend(
+            module.encoder.layer[num_layers_per_stage_accumulated[self.stage_manager.stage - 1] if self.stage_manager.
```
Comment (ver217) [inline colossalai/pipeline/policy/bert.py:267]: This is ugly. You can first `np.insert(num_layers_per_stage_accumulated, 0, 0)`, then
```python
start_idx = num_layers_per_stage_accumulated[self.stage_manager.stage - 1]
end_idx = num_layers_per_stage_accumulated[self.stage_manager.stage]
hold_layers.extend(module.encoder.layer[start_idx:end_idx])
```

File: colossalai/pipeline/policy/bert.py [hunk 2]
```diff
@@ -243,8 +251,9 @@ def custom_forward(*inputs):
 class BertModelPolicy(Policy):
 
     def __init__(self, stage_manager: PipelineStageManager, num_layers: int, num_stages: int):
+        super().__init__(stage_manager=stage_manager)
         self.stage_manager = stage_manager
-        self.layers_per_stage = self.distribute_layers(num_layers, num_stages)
+        self.layers_per_stage = super().distribute_layers(num_layers, num_stages)
```
Comment (ver217) [inline colossalai/pipeline/policy/bert.py:256]: You can call `self.distribute_layers()` directly.

File: colossalai/pipeline/policy/bert.py (hunk 1 above)
Comment (ver217) [reply colossalai/pipeline/policy/bert.py:267]: And this process can be extracted to base class.

Comment (CjhHa1) [reply colossalai/pipeline/policy/bert.py]: done

File: colossalai/pipeline/policy/base.py [hunk 3]
```diff
@@ -106,3 +108,20 @@ def parallelize_model(self, module: Module) -> Tuple[Dict[str, Parameter], Dict[
         self.replace_forward(module)
         shared_params = self.get_shared_params(module)
         return hold_params, hold_buffers, shared_params
+
+    def distribute_layers(self, num_layers: int, num_stages: int) -> List[int]:
```
[the remark this answers is not in the document]
Comment (CjhHa1) [reply colossalai/pipeline/policy/base.py]: done

File: colossalai/pipeline/policy/bert.py (hunk 2 above)
Comment (CjhHa1) [reply colossalai/pipeline/policy/bert.py]: done

File: colossalai/pipeline/policy/bert.py (hunk 1 above)
Comment (CjhHa1) [reply colossalai/pipeline/policy/bert.py]: extracted and also make it a static method

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**